include/envoy/http/filter.h

```diff
@@ -151,6 +151,19 @@ class StreamFilterCallbacks {
    */
   virtual Router::RouteConstSharedPtr route() PURE;
 
+  /**
+   * Invokes callback with a matched route, callback can choose to accept this route by returning
+   * Router::RouteMatchStatus::Accept or continue route match from last matched route by returning
+   * Router::RouteMatchStatus::Continue, if there are more routes available.
+   *
+   * Returns route accepted by the callback or nullptr if no match found or none of route is
+   * accepted by the callback.
+   *
+   * NOTE: clearRouteCache() must be called before invoking this method otherwise cached route will
+   * be returned directly to the caller and the callback will not be invoked.
+   */
+  virtual Router::RouteConstSharedPtr route(const Router::RouteCallback& cb) PURE;
```

Comment:
so this function is not guaranteed to be called by every call site, consider this case: 
1. in one filter you update route by offering a cb;
2. in a following filter it refresh the route, the picked special cluster is overrides by the no-cb return value

is this behavior desired?

Reply:
That's correct cb is valid for a given call site only, and it is local to a given filter.

Reply:
So this is test-only purpose change?  Would it be possible to test your use case with something not changing the core lib?

Reply:
It is not for test purpose, Let's say there are multiple routes that can be matched against the incoming request Envoy currently always choose first matching route. However in certain scenarios a custom filter wants, (may be we are experimenting with a L7 rule but not want to use it right away may be expose the selection in logs so the impact can be learned) Envoy to choose a route other than the first matching route, in such case filter will clear cached route and invoke route with a call back so it can ask Envoy to continue evaluating the route beyond the first match.

Reply:
so what happens after such evaluation? do you expect to keep this new route-picking behavior cross filters or just a one shot thing if you intend to only change route in this one specific filter?
I'd suggest you "latch" the callback to the stream if the new-route picking behavior is supposed not to be "reverted" by other filters down the chain which also refreshes  the cache.

Reply:
Once one filter decide to override the route selection and finally end up accepting one, it will be cached by Envoy and used by router filter for routing traffic, however any intermediate filter can again clear the cached route and invoke route, this is consistent with existing behavior.

Reply:
I understand that this "keep-on-route-selection-in-this-filter" is a one time thing in your use case, but it also introduce behavior inconsistency into the filter-chain w.r.t route-selection, which could be confusing in query processing. 
For example, an empty route entry could possibly trigger a new route selection whenever a filter requests clusterInfo from the decoder_calbacks_, in which case, it may return something totally different than the "override-filter" picked. 
would it be plausible if we set the callback w/ the stream and keep the consistency along the filter-chain?

Reply:
@stevenzzzz this is exactly how it works. It is consistent with the non-callback method which does exactly that. Once a route has been selected, it will stick.

Reply:
let's say the CB_ returns a nullptr for route_entry, skipped the non-null route, in the following refreshCachedRoute() call, the no-cb call will update the route_entry to a non-null roue-entry, erase the result made by the cb_, which introduce the inconsistency. 
I guess it's fine as  user gets to define their cb and configure their filter-chain.

Reply:
Yeah I see what you are saying. Unfortunately I don't see any easy answer here, because further filters that may have changed the route table may not know anything about the callback. IMO this is something that we have to live with, unless you can think of any clean solutions? IMO we should ship this and iterate as this is much better than what we have today.